# Analyzing Data for Heroes of Pymoli Using Pandas

In [57]:
# Dependencies / Importing the library we are using
import pandas as pd

# File to Load
file = "./Resources/purchase_data.csv"

# Read the purchase data file
df_purchase_data = pd.read_csv(file)
df_purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

### Player Count

In [58]:
# Count the total number of players
player = df_purchase_data["SN"].unique()
# list(player)

player_count = len(player)
# print(player_count)

# Print total player count in data frame
player_count_df = pd.DataFrame({"Total Players": [player_count]})
player_count_df

Total Players
0            576

### Purchasing Analysis (Total)

In [59]:
# Count the total number of unique items
item = df_purchase_data["Item ID"].unique()
# list(item)

item_count = len(item)
print(item_count)

# Calculate the average purchase price
avg_price = df_purchase_data["Price"].mean()
print(avg_price)

# Count the total number of purchases
purchase_count = len(df_purchase_data["Purchase ID"])
print(purchase_count)

# Calculate total revenue
revenue = df_purchase_data["Price"].sum()
print(revenue)

183
3.050987179487176
780
2379.77


In [60]:
# Print purchasing analysis (total) to a summary data frame
summary_purchasing_df = pd.DataFrame({"Number of Unique Items": [item_count],
                                      "Average Price": [avg_price],
                                      "Number of Purchases": [purchase_count],
                                      "Total Revenue": [revenue]})
summary_purchasing_df["Average Price"] = summary_purchasing_df["Average Price"].map("${:,.2f}".format)
summary_purchasing_df["Number of Purchases"] = summary_purchasing_df["Number of Purchases"].map("{:,}".format)
summary_purchasing_df["Total Revenue"] = summary_purchasing_df["Total Revenue"].map("${:,.2f}".format)
summary_purchasing_df = summary_purchasing_df.loc[:, ["Number of Unique Items", "Average Price", "Number of Purchases", "Total Revenue"]]
summary_purchasing_df

Number of Unique Items Average Price Number of Purchases Total Revenue
0                     183         $3.05                 780     $2,379.77

### Gender Demographics

In [61]:
# Calculate the count and percentage of players for each genders
player_by_gender = df_purchase_data.groupby("Gender")["SN"].nunique()
# print(gender_count)

player_by_gender_df = pd.DataFrame({"Total Count": player_by_gender})

player_percentage = (player_by_gender_df["Total Count"] / player_by_gender_df["Total Count"].sum()) * 100
player_by_gender_df["Percentage of Players"] = player_percentage
player_by_gender_df["Percentage of Players"] = player_by_gender_df["Percentage of Players"].map("{:.2f}%".format)
player_by_gender_df

Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%

### Purchasing Analysis (Gender)

In [62]:
# Calculate the count of purchase ID by each genders
purchase_by_gender = df_purchase_data.groupby("Gender")["Purchase ID"].count()
purchase_by_gender

# Print the count of purchase ID in data frame
purchase_by_gender_df = pd.DataFrame({"Purchase Count": purchase_by_gender})
# purchase_by_gender_df

# Calculate the average of purchase price by each genders
avg_purchase_by_gender = df_purchase_data.groupby("Gender")["Price"].mean()
purchase_by_gender_df["Average Purchase Price"] = avg_purchase_by_gender.map("${:.2f}".format)
# purchase_by_gender_df

# Calculate the total purchase price by each genders
total_price_by_gender = df_purchase_data.groupby("Gender")["Price"].sum()
purchase_by_gender_df["Total Purchase Value"] = total_price_by_gender.map("${:,.2f}".format)
# purchase_by_gender_df

In [63]:
# Calculate the average total purchase per person by each genders
# First, we calculate the total purchase per person
purchase_per_person = df_purchase_data.groupby("SN")["Price"].sum()
purchase_per_person_df = pd.DataFrame({"Total Purchase Per Person": purchase_per_person})
# purchase_per_person_df

# Second, we map the gender for each person
merge_gender = purchase_per_person_df.merge(df_purchase_data[["SN", "Gender"]], on = "SN")
merge_gender_no_dup = merge_gender.drop_duplicates(subset="SN", keep="first")
# merge_gender_no_dup

# Third, we calculate the average total purchase per person by each genders
purchase_per_person_by_gender = merge_gender_no_dup.groupby("Gender")["Total Purchase Per Person"].mean()
# purchase_per_person_by_gender

# Print to the summary table
purchase_by_gender_df["Average Total Purchase per Person"] = purchase_per_person_by_gender.map("${:.2f}".format)
purchase_by_gender_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Total Purchase per Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

### Age Demographics

In [64]:
df_purchase_data["Age"].max()

45

In [65]:
# Assign the data to different bins depending on 4 years age range
# Create the bins in which the data will be held
age_bins = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 50]

# Create the names for the bins
age_group_name = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Assign the age group to the data
df_purchase_data["Age Group"] = pd.cut(df_purchase_data["Age"], age_bins, labels = age_group_name)
df_purchase_data

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male      101   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Group  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [66]:
# Group the data frame by age group
age_group = df_purchase_data.groupby("Age Group")

# Count the number of players by age group
player_by_age_group = df_purchase_data.groupby("Age Group")["SN"].nunique()
# player_by_age_group

player_by_age_group_df = pd.DataFrame({"Total Count": player_by_age_group})
player_by_age_group_df

# Calculate the percentage of players by age group
player_percentage_by_age_group = (player_by_age_group_df["Total Count"] / player_by_age_group_df["Total Count"].sum()) * 100
player_by_age_group_df["Percentage of Players"] = player_percentage_by_age_group.map("{:.2f}%".format)
player_by_age_group_df

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

In [67]:
# Group the data frame by age group
age_group = df_purchase_data.groupby("Age Group")
# age_group

# Count the number of purchase count by age group
purchase_by_age = age_group["Purchase ID"].count()
# purchase_by_age

# Calculate the average of purchase price by age group
avg_purchase_by_age = age_group["Price"].mean()
# avg_purchase_by_age

# Calculate the total purchase price by age group
total_price_by_age = age_group["Price"].sum()
# total_price_by_age

# Print the results in summary table by age group
age_group_df = pd.DataFrame({"Purchase Count": purchase_by_age,
                            "Average Purchase Price": avg_purchase_by_age,
                            "Total Purchase Value": total_price_by_age})

In [68]:
# Calculating averange total purchase per person
purchase_per_person_by_age_group = age_group_df["Total Purchase Value"] / player_by_age_group_df["Total Count"]
age_group_df["Average Total Purchase per Person"] = purchase_per_person_by_age_group
age_group_df = age_group_df.loc[:,["Purchase Count", "Average Purchase Price",
                                  "Total Purchase Value", "Average Total Purchase per Person"]]
age_group_df["Purchase Count"] = age_group_df["Purchase Count"].map("{:,}".format)
age_group_df["Average Purchase Price"] = age_group_df["Average Purchase Price"].map("${:,.2f}".format)
age_group_df["Total Purchase Value"] = age_group_df["Total Purchase Value"].map("${:,.2f}".format)
age_group_df["Average Total Purchase per Person"] = age_group_df["Average Total Purchase per Person"].map("${:,.2f}".format)
age_group_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                              
<10                   23                  $3.35               $77.13   
10-14                 28                  $2.96               $82.78   
15-19                136                  $3.04              $412.89   
20-24                365                  $3.05            $1,114.06   
25-29                101                  $2.90              $293.00   
30-34                 73                  $2.93              $214.00   
35-39                 41                  $3.60              $147.67   
40+                   13                  $2.94               $38.24   

          Average Total Purchase per Person  
Age Group                                    
<10                                   $4.54  
10-14                                 $3.76  
15-19                                 $3.86  
20-24                                 $4.32  
25-29                                 $3.81  
30-34                                 $4.12  
35-39                                 $4.76  
40+                                   $3.19

### Top Spenders

In [77]:
# Calculate the total purchase value for each person
total_purchase_per_person = df_purchase_data.groupby("SN")["Price"].sum()

# Calculate the average purchase price for each person
avg_purchase_per_person = df_purchase_data.groupby("SN")["Price"].mean()

# Count the number of purchase ID per person
count_purchase_per_person = df_purchase_data.groupby("SN")["Purchase ID"].count()

purchase_per_person_df = pd.DataFrame({"Purchase Count": count_purchase_per_person,
                                      "Average Purchase Price": avg_purchase_per_person,
                                      "Total Purchase Value": total_purchase_per_person})

# Sort the data frame by the highest total purchase price and format values
sort_by_total_price = purchase_per_person_df.sort_values("Total Purchase Value", ascending = False)
sort_by_total_price["Average Purchase Price"] = sort_by_total_price["Average Purchase Price"].map("${:,.2f}".format)
sort_by_total_price["Total Purchase Value"] = sort_by_total_price["Total Purchase Value"].map("${:,.2f}".format)
sort_by_total_price = sort_by_total_price.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value"]].head(5)
sort_by_total_price




Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Most Popular Items

In [79]:
# Group data frame by Item ID and Item Name
item_groups = df_purchase_data.groupby(["Item ID", "Item Name"])

# Count the number of purchase by Item ID and Name
purchase_by_item = item_groups["Purchase ID"].count()
purchase_by_item_df = pd.DataFrame({"Purchase Count": purchase_by_item})
# purchase_by_item_df

# Add item price to the data
merge_item_price = purchase_by_item_df.merge(df_purchase_data[["Item ID", "Item Name", "Price"]], on = "Item Name")
merge_item_price_no_dup = merge_item_price.drop_duplicates(subset = "Item Name", keep = "first")
# merge_item_price_no_dup

# Group Item ID and Item Name with individual item price
item_price = merge_item_price_no_dup.groupby(["Item ID", "Item Name"])["Price"].sum()

# Calculate the total purchase value by Item ID and Name
total_price_by_item = item_groups["Price"].sum()
# total_price_by_item

# Print in summary table
purchase_by_item_df["Item Price"] = item_price
purchase_by_item_df["Total Purchase Value"] = total_price_by_item
# purchase_by_item_df

# Sort the data frame by the 5 most popular items by purchase count and format the values
sort_by_purchase_count = purchase_by_item_df.nlargest(5, "Purchase Count")
sort_by_purchase_count["Item Price"] = sort_by_purchase_count["Item Price"].map("${:,.2f}".format)
sort_by_purchase_count["Total Purchase Value"] = sort_by_purchase_count["Total Purchase Value"].map("${:,.2f}".format)
sort_by_purchase_count = sort_by_purchase_count.loc[:, ["Purchase Count", "Item Price", "Total Purchase Value"]]
sort_by_purchase_count

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
108     Extraction, Quickblade Of Trembling Hands                  9   
145     Fiery Glass Crusader                                       9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
145     Fiery Glass Crusader                              $4.58   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
145     Fiery Glass Crusader                                       $41.22  
19      Pursuit, Cudgel of Necromancy                               $8.16

### Most Profitable Items

In [80]:
# Recall purchase by item data frame, and convert total purchase value back to float
# purchase_by_item_df["Total Purchase Value"] = total_price_by_item
# purchase_by_item_df

# Sort the summary data of 'Purchase by Item' by total purchase value and format the values
sort_by_total_purchase = purchase_by_item_df.nlargest(5, "Total Purchase Value")
sort_by_total_purchase["Item Price"] = sort_by_total_purchase["Item Price"].map("${:,.2f}".format)
sort_by_total_purchase["Total Purchase Value"] = sort_by_total_purchase["Total Purchase Value"].map("${:,.2f}".format)
sort_by_total_purchase = sort_by_total_purchase.loc[:, ["Purchase Count", "Item Price", "Total Purchase Value"]]
sort_by_total_purchase


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

### Summary Analysis

- 84% of all players are male and male purchase more in total compared with other gender. However, if we see the average purchase per person per gender, male purchased the least compared with other gender per person.
- In terms of age group, the highest number of players are within the range of 20-24 years old (258 out of 576 players). This age group also has the highest number of purchase count (365 out of 780 purchases).
- The player "Lisosia93" spent the most out of all players (there are a total of 576 players).
- Item ID "178", Item Name "Oathbreaker, Last Hope of the Breaking Storm" was the most popular and most profitable items compared with other items (there are 183 unique items).